[bpo-28414] Make all hostnames in SSL module IDN A-labels

tiran opened the pull request:
Alternative approach to PR #3010

- Really all server_hostname attributes are IDN A-labels
- SSLContext.set_servername_callback argument is IDN A-label, too
- Replace all externally replaced certs with proper certs from internal test helper

<!-- issue-number: bpo-28414 -->
https://bugs.python.org/issue28414
<!-- /issue-number -->

njsmith committed:
[bpo-28414] In SSL module, store server_hostname as an A-label

Historically, our handling of international domain names (IDNs) in the
ssl module has been very broken. The flow went like:

1. User passes server_hostname= to the SSLSocket/SSLObject
   constructor. This gets normalized to an A-label by using the
   PyArg_Parse "et" mode: bytes objects get passed through
   unchanged (assumed to already be A-labels); str objects get run
   through .encode("idna") to convert them into A-labels.

2. newPySSLSocket takes this A-label, and for some reason decodes
   it *back* to a U-label, and stores that as the object's
   server_hostname attribute.

3. Later, this U-label server_hostname attribute gets passed to
   match_hostname, to compare against the hostname seen in the
   certificate. But certificates contain A-labels, and match_hostname
   expects to be passed an A-label, so this doesn't work at all.

This PR fixes the problem by removing the pointless decoding at step
2, so that internally we always use A-labels, which matches how
internet protocols are designed in general: A-labels are used
everywhere internally and on-the-wire, and U-labels are basically just
for user interfaces.

This also matches the general advice to handle encoding/decoding once
at the edges, though for backwards-compatibility we continue to use
'str' objects to store A-labels, even though they're now always
ASCII. Technically there is a minor compatibility break here: if a
user examines the .server_hostname attribute of an ssl-wrapped socket,
then previously they would have seen a U-label like "pythön.org", and
now they'll see an A-label like "xn--pythn-mua.org". But this only
affects non-ASCII domain names, which have never worked in the first
place, so it seems unlikely that anyone is relying on the old
behavior.

This PR also adds an end-to-end test for IDN hostname
validation. Previously there were no tests for this functionality.

Fixes bpo-28414.

tiran committed:
Drop trustme certs

All test certs must be generated by CPython's own test helper.

Signed-off-by: Christian Heimes <christian@python.org>

tiran committed:
Fix IDN SAN handling

Signed-off-by: Christian Heimes <christian@python.org>

alex commented on Doc/whatsnew/3.7.rst:
```diff
@@ -441,6 +441,19 @@ can be set within the scope of a group.
+expected hostname in A-label form (``"xn--pythn-mua.org"``), rather
+than the U-label form (``"pythön.org"``). The server name argument to
+:meth:`ssl.SSLContext.set_servername_callback`` is an A-label instead of
+an U-labe, too. (Contributed by Nathaniel J. Smith and Christian Heimes
```
"labe"

alex replied:
(also not sure what the deal with github rendering all this with underlines is; potential reST issue?)

tiran replied:
It's caused by a supernumerous backtick. I have fixed both issues.

alex commented on Lib/test/test_asyncio/test_events.py:
```diff
@@ -1181,6 +1181,8 @@ def test_create_unix_server_ssl_verified(self):
         server.close()
         self.loop.run_until_complete(proto.done)
 
+    maxDiff = None
```
Unrelated?

tiran replied:
Remnants of debugging failing tests

njsmith commented on Lib/ssl.py:
```diff
@@ -389,6 +400,20 @@ def set_npn_protocols(self, npn_protocols):
+            self.sni_callback = None
+        else:
+            if not hasattr(server_name_callback, '__call__'):
+                raise TypeError("not a callable object")
```
Why not `if not callable(server_name_callback)`?

tiran replied:
Ah right, ``callable`` was re-added in Python 3.2. I'll change it.

njsmith commented on Lib/ssl.py:
```diff
@@ -389,6 +400,20 @@ def set_npn_protocols(self, npn_protocols):
+
+            def shim_cb(sslobj, servername, sslctx):
+                if servername is not None:
+                    servername = servername.encode("ascii").decode("idna")
```
It would be simpler to use `_encode_hostname` here, no?

tiran replied:
I don't mind to use the method here, too.

njsmith commented on Lib/ssl.py:
```diff
@@ -355,13 +355,20 @@ def __new__(cls, protocol=PROTOCOL_TLS, *args, **kwargs):
 
-    def __init__(self, protocol=PROTOCOL_TLS):
-        self.protocol = protocol
+    def _encode_hostname(self, hostname):
```
Maybe make this a top-level utility function, instead of a method?

tiran replied:
I'd prefer to keep this a method. It makes it easier to modify the behavior in subclasses or in tests.

njsmith committed:
Small doc updates

njsmith commented on Modules/_ssl.c:
```diff
@@ -4217,8 +4248,12 @@ static PyGetSetDef context_getsetlist[] = {
     {"options", (getter) get_options,
                 (setter) set_options, NULL},
+    {"protocol", (getter) get_protocol,
+                 NULL, NULL},
```
Why is this new `protocol` attribute in this PR? Should it be documented or anything?

tiran replied:
It's not new: https://docs.python.org/3/library/ssl.html#ssl.SSLContext.protocol

I just moved it from Python code into C code so I can access the value from C methods. The new setter for the callback uses the value to prevent setting a SNI callback for a client-only context.

njsmith replied:
Ah, gotcha.

njsmith commented on Modules/_ssl.c:
```diff
@@ -4005,45 +4009,62 @@ _servername_callback(SSL *s, int *al, void *args)
+                Py_INCREF(result);
+                Py_SETREF(ssl->ctx, (PySSLContext *)result);
+                SSL_set_SSL_CTX(ssl->ssl, ssl->ctx->ctx);
+            }
```
The docs just say that `sni_callback` has to return None. What's this extra context setting stuff about? Should it be removed?

tiran replied:
I added the code for future compatibility with PEP 543. Rather than modifying socket.context, the callback can return a new context and let the API take care of the rest.

By the way it doesn't have to be None. Other values signal an error:

> The server_name_callback function must return None to allow the TLS negotiation to continue. If a TLS failure is required, a constant ALERT_DESCRIPTION_* can be returned. Other return values will result in a TLS fatal error with ALERT_DESCRIPTION_INTERNAL_ERROR.

njsmith replied:
Ok yeah, I think adding new undocumented features in b2 for partial compatibility with an unfinished PEP is probably a bad plan, so thanks for removing that :-). It'll be easy to add back later.

njsmith committed:
Update blurb to match what this PR does now

njsmith reviewed (commented):
I pushed some prose changes, and here are some comments on the code. Overall looks good, but there are a few small cleanups to consider, and a few points where I have questions.

tiran committed:
Address review comments

Drop extra code for PEP 543 future compatibility in sni callback

Use callable() and encode_hostname in shim for old SNI callback.

Signed-off-by: Christian Heimes <christian@python.org>
